serverless/serverless
# Allow Custom Per-Function Roles

## What did you implement:

Closes Issue[s]: #1807 and #1895 by allowing per-function role declarations with fallback to provider or default policy/role


## How did you implement it:



Allow each function to declare the ARN of the role that it is to execute under.
If any function has no role ARN but provider role ARN is defined then use that role for such functions.
If any function does not have a specified role (even by falling back to a provider-wide role) then add the default policy and role so that it can be used for such functions.
Break out the portions of the function assignment to CFT logic into their discrete units so that the registered plugin code is a readable summary.  Add comments to the discrete units.
## How can we verify it:



Added tests that check that the default role and policy are not added if every function has a role.
Added tests that check that every function that has a declared role gets it assigned.
Added tests that check that every function that has no declared role but where a provider role is declared gets the provider role assigned.
## Todos:
